internal/app/synchronizer/synchronizer_service.go

```diff
@@ -46,6 +132,165 @@ func (s *synchronizerService) Register(ctx context.Context, req *pb.RegisterRequ
 // matches accepted by it are returned as results.
-func (s *synchronizerService) EvaluateProposals(ctx context.Context, req *pb.EvaluateProposalsRequest) (*pb.EvaluateProposalsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *synchronizerService) EvaluateProposals(ctx context.Context, req *ipb.EvaluateProposalsRequest) (*ipb.EvaluateProposalsResponse, error) {
+	synchronizerServiceLogger.Debugf("Request to evaluate propsals - id:%v, proposals: %v", req.Id, getMatchIds(req.Match))
+	// pendingRequests keeps track of number of requests pending. This is incremented
+	// in addProposals while holding the state mutex. The count should be decremented
+	// only after this request completes.
+	err := s.addProposals(req.Id, req.Match)
+	if err != nil {
+		return nil, err
+	}
+
+	// If proposals were successfully added, then the pending request count has been
+	// incremented and should only be reduced when this request is completed.
+	defer s.cycleData.pendingRequests.Done()
+
+	// fetchResults blocks till results are available. It does not change the
+	// state of the synchronizer itself (since multiple concurrent of these can
+	// be in progress). The evaluation routine tracks the completion of all of
+	// these and then changes the state of the synchronizer to idle.
+	results := s.fetchResults(req.Id)
+	return &ipb.EvaluateProposalsResponse{Match: results}, nil
+}
+
+func (s *synchronizerService) addProposals(id string, proposals []*pb.Match) error {
+	s.stateMutex.Lock()
+	defer s.stateMutex.Unlock()
+
+	if !s.proposalCollection() {
+		return status.Error(codes.DeadlineExceeded, "synchronizer currently not accepting match proposals")
+	}
+
+	if _, ok := s.cycleData.requests[id]; !ok {
+		return status.Error(codes.DeadlineExceeded, "request not present in the current synchronization cycle")
+	}
+
+	if len(s.cycleData.requests[id].proposals) > 0 {
+		// We do not currently support submitting multiple separate proposals for a registered request.
+		return status.Error(codes.Internal, "multiple proposal addition for a request not supported")
+	}
+
+	// Proposal can be added for this request so increment the pending requests
+	s.cycleData.pendingRequests.Add(1)
+
+	// Copy the proposals being added to the request data for the specified id.
+	s.cycleData.requests[id].proposals = make([]*pb.Match, len(proposals))
+	copy(s.cycleData.requests[id].proposals, proposals)
+	return nil
+}
+
+func (s *synchronizerService) proposalCollection() bool {
+	return s.state == proposalCollection || s.state == requestRegistration
+}
+
+func (s *synchronizerService) fetchResults(id string) []*pb.Match {
+	t := time.Now()
+	var results []*pb.Match
+
+	// fetchMatches will block till evaluation is in progress. The resultsReady channel
+	// is closed only when evaluation is complete. This is used to signal all waiting
+	// requests that the results are ready.
+	<-s.cycleData.resultsReady
```

Comment:
Add a TODO to support timeout here. I'll add a primitive to handle this. Basically I don't think we should ever wait for channels like this in the serving path. It's prone to denial of service issues.

Also file an issue against me to add it.

Reply:
Good point - yes, if the user evaluation blocks this .. it would queue up new registrations.. Added a TODO to timeout here - assigned issue to you. Thanks!